### PR TITLE
Increase Test Coverage for UserViewModel and UserRepositoryFireStore

### PR DESCRIPTION
**Pull request by nada-ezziti**
# Increase Test Coverage for UserViewModel and UserRepositoryFireStore

## Description

The goal of this pull request is to increase the test coverage for `UserViewModel` and `UserRepositoryFireStore`. The initial line coverage for both classes was approximately **60%**. By adding extensive tests, particularly focusing on failure cases and edge scenarios.

### UserViewModelTest

- **Added Tests for Success Cases:**
  - Verified that the `MutableStateFlow` properties are correctly updated on successful repository callbacks.
  - Ensured methods like `getUserById`, `getUserName`, `getProfilePictureUrl`, and `getOngoingChallenges` handle successful responses appropriately.

- **Added Tests for Failure Cases:**
  - Simulated exceptions and repository failures to test how the `UserViewModel` handles errors.
  - Verified that error states are correctly set and internal states remain consistent when failures occur.

### UserRepositoryFireStoreTest

- **Enhanced Failure Case Coverage:**
  - Added tests to simulate failures in Firestore operations such as `get()`, `update()`, and `set()`.
  - Tested how the repository handles null or non-existent documents and fields.

- **Indirect Testing of Private Methods:**
  - By thoroughly testing public methods, we indirectly covered private helper functions, increasing overall line coverage.

- **Tested Edge Cases and Exception Handling:**
  - Included tests for scenarios like null values, empty lists, and exceptions thrown in asynchronous callbacks.
  - Ensured that all branches and code paths are exercised in the tests.

_______________

## Notes

- **Dependencies:**
  - Added necessary testing dependencies such as `androidx.arch.core:core-testing` for `InstantTaskExecutorRule`.
  - Ensured that all mocks and test utilities are properly initialized and cleaned up to prevent interference between tests.

________________

## Conclusion

By adding these comprehensive tests, we've strengthened the foundation of our codebase, ensuring better stability and easier maintenance moving forward. The focus on failure cases and edge scenarios provides confidence that the application can handle unexpected situations gracefully.

### CODE REVIEW

**Comment by nada-ezziti** [comment]
> I'm having a hard time understanding what `InstantTaskExecutorRule` is used for, and why your new dependency `androidx.arch.core:core-testing` is necessary. Could you help us with some insight on what got you to add this dependency?

InstantTaskExecutorRule ensures that LiveData and other architecture components execute their tasks synchronously during tests, making it easier to test ViewModels and LiveData without concurrency issues. The androidx.arch.core:core-testing dependency provides this rules. It simplifies and stabilizes tests involving LiveData and ViewModels.

File: app/src/test/java/com/github/se/signify/model/user/UserViewModelTest.kt [hunk 1]
```diff
@@ -120,4 +141,380 @@ class UserViewModelTest {
...
+          null
+        }
+        .whenever(userRepository)
+        .setInitialQuestAccessDate(eq(currentUserId), any(), any(), any())
+
+    // Act
+    userViewModel.checkAndUnlockNextQuest()
+
+    // Assert
+    assertEquals("1", userViewModel.unlockedQuests.value)
```
Comment (samymessouak) [inline app/src/test/java/com/github/se/signify/model/user/UserViewModelTest.kt:456]: Adding comments explaining the purpose of each assertion and the expected state after each action can make the tests easier to follow for other team members.

**Review by samymessouak — CHANGES_REQUESTED** [review]
Overall, this pull request significantly improves the test coverage and robustness of UserViewModel and UserRepositoryFireStore. The existing changes already add value by covering a variety of success and failure scenarios, and they set a strong foundation for ensuring stability. However, implementing the suggested changes will further enhance the readability, maintainability, and consistency of the test suite. These small improvements will make it easier for all team members to understand, extend, and maintain the tests in the future, ultimately contributing to a more resilient and well-documented codebase.

File: app/src/test/java/com/github/se/signify/model/user/UserViewModelTest.kt [hunk 2]
```diff
@@ -120,4 +141,380 @@ class UserViewModelTest {
...
+        .whenever(userRepository)
+        .getFriendsList(eq(currentUserId), any(), any())
+
+    // Act
+    userViewModel.getFriendsList()
+
+    // Assert
+    // Error is logged; verify the method call
+    verify(userRepository).getFriendsList(eq(currentUserId), any(), any())
+  }
```
[the remark this answers is not in the document]
Comment (nada-ezziti) [reply app/src/test/java/com/github/se/signify/model/user/UserViewModelTest.kt:182]: Thanks for your review, I just did a commit adding your suggestion.

**Review by samymessouak — APPROVED** [review]
Perfect changes, makes it clearer for me and probably for the others too. Thank you Nada
